modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -259,38 +265,136 @@ protected void doStop() {
     @Override
     protected void doClose() {}
 
-    private static final class SuccessfulRequestTracker implements HttpPipelinePolicy {
-        private static final Logger logger = LogManager.getLogger(SuccessfulRequestTracker.class);
+    static class RequestMetrics {
+        private volatile long timeToResponseInMillis = -1;
+        private volatile int requestCount;
+        private volatile int errorCount;
+        private volatile int throttleCount;
+        private volatile int statusCode;
+
+        int getRequestCount() {
+            return requestCount;
+        }
+
+        int getErrorCount() {
+            return errorCount;
+        }
+
+        int getStatusCode() {
+            return statusCode;
+        }
+
+        int getThrottleCount() {
+            return throttleCount;
+        }
+
+        /**
+         * Time to response, or -1 if response was never received
+         */
+        long getTimeToResponseInMillis() {
+            return timeToResponseInMillis;
+        }
+
+        @Override
+        public String toString() {
+            return "RequestMetrics{"
+                + "timeToResponseInMillis="
+                + timeToResponseInMillis
+                + ", requestCount="
+                + requestCount
+                + ", errorCount="
+                + errorCount
+                + ", throttleCount="
+                + throttleCount
+                + ", statusCode="
+                + statusCode
+                + '}';
+        }
+    }
+
+    private enum RetryMetricsTracker implements HttpPipelinePolicy {
+        INSTANCE;
+
+        @Override
+        public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+            Optional<Object> metricsData = context.getData(RequestMetricsTracker.ES_REQUEST_METRICS_CONTEXT_KEY);
+            if (metricsData.isPresent() == false) {
+                assert false : "No metrics object associated with request " + context.getHttpRequest();
+                return next.process();
+            }
```

Comment:
This would be a bug if this handler were in the pipeline and the metrics wasn't populated in the context, but I made it an assertion because it should be catchable with tests.